doubly_linked_list/doubly_linked_list.py

```diff
@@ -1,64 +1,139 @@
+        current_prev = self.prev
+        self.prev = ListNode(value, current_prev, self)
+        if current_prev:
+            current_prev.next = self.prev
+
+    """Rearranges this ListNode's previous and next pointers
   accordingly, effectively deleting this ListNode."""
-  def delete(self):
-    if self.prev:
-      self.prev.next = self.next
-    if self.next:
-      self.next.prev = self.prev
+
+    def delete(self):
+        if self.prev:
+            self.prev.next = self.next
+        if self.next:
+            self.next.prev = self.prev
+
 
 """Our doubly-linked list class. It holds references to
 the list's head and tail nodes."""
+
+
 class DoublyLinkedList:
-  def __init__(self, node=None):
-    self.head = node
-    self.tail = node
+    def __init__(self, node=None):
+        self.head = node
+        self.tail = node
+
+    def add_to_head(self, value):
+        if self.head:
+            self.head.insert_before(value)
+            self.head = self.head.prev
+        else:
+            self.head = ListNode(value)
+            self.tail = self.head
 
-  def add_to_head(self, value):
-    pass
+    def remove_from_head(self):
+        if self.head:
+            new_head = self.head.next
+            temp = self.head.value
+            self.head.delete()
+            if not self.head:
+                self.tail = None
+            return temp
+        else:
+            return None
 
-  def remove_from_head(self):
-    pass
+    def add_to_tail(self, value):
+        if self.tail:
+            self.tail.insert_after(value)
+            self.tail = self.tail.next
+        else:
+            self.tail = ListNode(value)
+            self.head = self.tail
 
-  def add_to_tail(self, value):
-    pass
+    def remove_from_tail(self):
+        if self.tail:
+            new_tail = self.tail.prev
+            temp = self.tail.value
+            self.tail.delete()
+            self.tail = new_tail
+            if not self.tail:
+                self.head = None
+            return temp
+        else:
+            return None
 
-  def remove_from_tail(self):
-    pass
+    def move_to_front(self, node):
+        if node.prev:
+            node.prev.next = node.next
+        if node.next:
+            node.next.prev = node.prev
+        node.prev = None
+        if self.head:
+            self.head.prev = node
+            node.next = self.head
+            self.head = node
+        else:
+            self.head = node
+            self.tail = node
 
-  def move_to_front(self, node):
-    pass
+    def move_to_end(self, node):
+        if node.prev:
+            node.prev.next = node.next
+        if node.next:
+            node.next.prev = node.prev
+        node.next = None
+        if self.tail:
+            self.tail.next = node
+            node.prev = self.tail
+            self.tail = node
+        else:
+            self.head = node
+            self.tail = node
 
-  def move_to_end(self, node):
-    pass
+    def delete(self, node):
```

Comment:
Try to use the codes you are provided with out-of-the-box instead to adding new parameters to existing functions.  With the original code, you can simply do:  node_to_delete.delete() , instead of doing: delete(node_to_delete)